Add ViViT

jegork opened the pull request:
# What does this PR do?

Fixes #15666

Add Video Vision Transformer to transformers. This PR implements a spacetime version of the Video Vision Transformer from the original paper. 

I have provided the model weights here https://huggingface.co/jegormeister/vivit-b-16x2-kinetics400 
I will try to add Factorised Encoder version later on (these are the two versions that authors provide weight for).


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? https://github.com/huggingface/transformers/issues/15666
- [x] Did you make sure to update the documentation with your changes? I have added the documentation, but I have troubles testing it as I couldn't run the preview command of the doc-builder, so if someone has the possibility to run and check it, I will be really grateful!
- [x] Did you write any new necessary tests? WIP


## Who can review?

@LysandreJik answered to the original issue so would be great if you could assist with the PR or suggesting who could. Thanks!

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- albert, bert, xlm: @LysandreJik
- blenderbot, bart, marian, pegasus, encoderdecoder,  t5: @patrickvonplaten, @patil-suraj
- longformer, reformer, transfoxl, xlnet: @patrickvonplaten
- fsmt: @stas00
- funnel: @sgugger
- gpt2: @patrickvonplaten, @LysandreJik
- rag: @patrickvonplaten, @lhoestq
- tensorflow: @LysandreJik

Library:

- benchmarks: @patrickvonplaten
- deepspeed: @stas00
- ray/raytune: @richardliaw, @amogkam
- text generation: @patrickvonplaten
- tokenizers: @n1t0, @LysandreJik
- trainer: @sgugger
- pipelines: @LysandreJik

Documentation: @sgugger

HF projects:

- datasets: [different repo](https://github.com/huggingface/datasets)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Examples:

- maintained examples (not research project or legacy): @sgugger, @patil-suraj
- research_projects/bert-loses-patience: @JetRunner
- research_projects/distillation: @VictorSanh

 -->

jegork committed:
Add model

jegork committed:
Add ability to get classification head weights

jegork committed:
Add docs

jegork committed:
Add imports to __init__.py

jegork committed:
Run style

fcakyon commented:
Hello, @jegork thanks for the PR! Having some experience in adding tests and fixing repo-consistency/style, I can help you with these aspects if you need any :+1:  Feel free to tag me when needed.

sgugger commented:
cc @alaradirik and @NielsRogge

alaradirik commented:
Hi @jegork, thanks for working on this! 

I saw that it's not possible to import and test ViViT on your PR branch yet, this is because a lot of files need to be edited to properly  import the new modules (ViViTConfig, ViViTModel, etc.). You can refer to this [PR](https://github.com/huggingface/transformers/pull/20459) to see what files need to be edited. 

You can either make these changes manually or run the `transformers-cli add-new-model` command, which automatically takes care of a lot of these changes and initializes the model-specific files (modelling_vivit.py, etc.). You can learn more about this over [here.](https://github.com/huggingface/transformers/blob/main/docs/source/en/add_new_model.mdx)

Once you are done, you can run the `make fixup` command to make sure your code passes the [style, quality and repo consistency CI tests](https://huggingface.co/docs/transformers/contributing).

cc @sgugger @NielsRogge

jegork commented:
Hey @alaradirik, thanks for your reply and guidance! 

I will address what you suggested and add tests by the end of the week.

jegork committed:
Fix imports and add mdx doc

jegork committed:
Run style

NielsRogge commented:
@jegork really cool work, as a next step you could try to make the CI as green as possible? Currently there are many failing checks (10 failing and 9 successful). You can click on "details" -> "artifacts" -> "failures long" to see why exactly a check has failed.

You will also need to rebase on the main branch due to some issues with TF versions which are fixed on the upstream branch:
```
git remote add upstream https://github.com/huggingface/transformers.git
git fetch upstream
git rebase upstream/main
```

jegork commented:
@NielsRogge thanks for your reply! I was indeed looking at the CI and not picking up the issue because most of them display `No module named 'keras.saving.hdf5_format'` as the error, but it seems that is exactly what you mentioned - the recent update to the TF version in the main branch. I will look into that today!

NielsRogge commented on docs/source/en/model_doc/vivit.mdx:
```diff
@@ -0,0 +1,54 @@
+
+[[autodoc]] ViViTConfig
+
+## ViViTFeatureExtractor
```
There's no need to add a feature extractor anymore, we've renamed those to image processors

NielsRogge commented on src/transformers/__init__.py:
```diff
@@ -402,6 +402,12 @@
+    "models.vivit": [
+        "VIVIT_PRETRAINED_CONFIG_ARCHIVE_MAP",
+        "ViViTConfig",
+        "ViViTFeatureExtractor",
```
`FeatureExtractor` can be removed everywhere

NielsRogge commented on src/transformers/__init__.py:
```diff
@@ -849,6 +855,17 @@
+    _import_structure["models.vivit"].extend(
+        [
+            "VIVIT_PRETRAINED_MODEL_ARCHIVE_LIST",
+            "ViViTLayer",
```
You can remove ViViTLayer and load_tf_weights_in_vivit everywhere from the inits

NielsRogge commented on src/transformers/__init__.py:
```diff
@@ -3904,8 +3927,6 @@
             top_k_top_p_filtering,
         )
         from .modeling_utils import PreTrainedModel
-
-        # PyTorch model imports
```
Please don't remove these 2 lines

NielsRogge commented on src/transformers/__init__.py:
```diff
@@ -4963,6 +4984,16 @@
             ViTMSNModel,
             ViTMSNPreTrainedModel,
         )
+
+        # PyTorch model imports
```
```suggestion
```

NielsRogge commented on src/transformers/models/vivit/configuration_vivit.py:
```diff
@@ -0,0 +1,119 @@
+# coding=utf-8
+# Copyright 2022 Anurag Arnab, Mostafa Dehghani, Georg Heigold, Chen Sun, Mario Lučić, Cordelia Schmid and The HuggingFace Inc. team. All rights reserved.
```
```suggestion
# Copyright 2022 The HuggingFace Inc. team. All rights reserved.
```
We usually don't add the author names to the configuration

NielsRogge commented on src/transformers/models/vivit/configuration_vivit.py:
```diff
@@ -0,0 +1,119 @@
+
+class ViViTConfig(PretrainedConfig):
+    r"""
+    This is the configuration class to store the configuration of a [`~ViViTModel`]. It is used to instantiate an ViViT
```
```suggestion
    This is the configuration class to store the configuration of a [`ViViTModel`]. It is used to instantiate a ViViT
```

NielsRogge commented on src/transformers/utils/dummy_pt_objects.py:
```diff
@@ -5677,6 +5677,41 @@ def __init__(self, *args, **kwargs):
         requires_backends(self, ["torch"])
 
 
+VIVIT_PRETRAINED_MODEL_ARCHIVE_LIST = None
```
Please also add a test file under tests/models/vivit, see timesformer as an example

sgugger commented:
Also please use `Vivit` to prefix all the classes (instead of `ViViT`), so `VivitConfig`, `VivitModel` etc (liek `BertConfig`, `BertModel` etc.). Users have complained a lot about the casing we use for our models as it makes them harder to find in the lib :-)

jegork committed:
Fix copyright

jegork committed:
Fix config docstring

jegork committed:
Remove imports of ViViTLayer and load_tf_weights_in_vivit

jegork committed:
Remove FeatureExtractor and replace with ImageProcessor everywhere

jegork committed:
Remove ViViTForPreTraining from vivit.mdx

jegork committed:
Change ViViT -> Vivit everywhere

jegork commented:
@NielsRogge apparently I have run what you provided incorrectly, sorry for that!

jegork committed:
Add model_doc to _toctree.yml

jegork committed:
Replace tuples with lists in arguments of VivitConfig

jegork committed:
Rename patch_size to tubelet_size in TubeletEmbeddings

jegork committed:
Fix checkpoint names

jegork committed:
Add tests

jegork committed:
Remove unused num_frames

jegork committed:
Fix imports for VivitImageProcessor

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_20441). All of your documentation changes will be reflected on that endpoint.

jegork committed:
Minor fixes

jegork committed:
Decrease number of frames in VivitModelTester from 32 to 16

jegork committed:
Decrease number of frames in VivitModelTester from 16 to 8

jegork committed:
Add initialization for pos embeddings

jegork committed:
Rename Vivit -> ViViT in some places

jegork commented:
@NielsRogge would be great if you could take a look at the changes I made. The CI also seems to be all good. 
Thanks!

NielsRogge commented on docs/source/en/_toctree.yml:
```diff
@@ -452,6 +452,8 @@
       - local: model_doc/vit_msn
         title: ViTMSN
+      - local: model_doc/vivit
+        title: Vivit
```
```suggestion
        title: ViViT
```

NielsRogge commented on docs/source/en/model_doc/vivit.mdx:
```diff
@@ -0,0 +1,44 @@
+specific language governing permissions and limitations under the License.
+-->
+
+# Video Vision Transformer
```
```suggestion
# Video Vision Transformer (ViViT)
```

NielsRogge commented on docs/source/en/model_doc/vivit.mdx:
```diff
@@ -0,0 +1,44 @@
+
+## Overview
+
+The Vivit model was proposed in [ViViT: A Video Vision Transformer](https://arxiv.org/abs/2103.15691) by Anurag Arnab, Mostafa Dehghani, Georg Heigold, Chen Sun, Mario Lučić, Cordelia Schmid.
```
```suggestion
The ViViT model was proposed in [ViViT: A Video Vision Transformer](https://arxiv.org/abs/2103.15691) by Anurag Arnab, Mostafa Dehghani, Georg Heigold, Chen Sun, Mario Lučić, Cordelia Schmid.
```

NielsRogge commented on docs/source/en/model_doc/vivit.mdx:
```diff
@@ -0,0 +1,44 @@
+## Overview
+
+The Vivit model was proposed in [ViViT: A Video Vision Transformer](https://arxiv.org/abs/2103.15691) by Anurag Arnab, Mostafa Dehghani, Georg Heigold, Chen Sun, Mario Lučić, Cordelia Schmid.
+The paper proposes one of the first successful pure-transformer based set of models for video understanding.
```
```suggestion
The paper proposes one of the first successful pure Transformer-based set of models for video understanding.
```

NielsRogge commented on src/transformers/__init__.py:
```diff
@@ -779,6 +783,7 @@
     _import_structure["models.vilt"].extend(["ViltFeatureExtractor", "ViltImageProcessor", "ViltProcessor"])
     _import_structure["models.vit"].extend(["ViTFeatureExtractor", "ViTImageProcessor"])
     _import_structure["models.vit_hybrid"].extend(["ViTHybridImageProcessor"])
+    _import_structure["models.vivit"].extend(["VivitImageProcessor"])
```
```suggestion
    _import_structure["models.vivit"].append("VivitImageProcessor")
```
(nit)

NielsRogge commented on src/transformers/__init__.py:
```diff
@@ -890,6 +895,7 @@
     _import_structure["modeling_utils"] = ["PreTrainedModel"]
 
     # PyTorch models structure
+
```
Please don't touch this line

NielsRogge commented on src/transformers/__init__.py:
```diff
@@ -913,6 +919,7 @@
             "ASTPreTrainedModel",
         ]
     )
+
```
Please don't touch this line

NielsRogge commented on src/transformers/__init__.py:
```diff
@@ -5177,6 +5194,14 @@
             ViTMSNModel,
             ViTMSNPreTrainedModel,
         )
+
+        # PyTorch model imports
```
Please don't move the "PyTorch model imports" line, should be at the top of the first model (albert)

NielsRogge commented on src/transformers/models/vivit/configuration_vivit.py:
```diff
@@ -0,0 +1,119 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" Vivit model configuration"""
```
```suggestion
""" ViViT model configuration"""
```

NielsRogge commented on src/transformers/models/vivit/configuration_vivit.py:
```diff
@@ -0,0 +1,119 @@
+    Configuration objects inherit from [`PretrainedConfig`] and can be used to control the model outputs. Read the
+    documentation from [`PretrainedConfig`] for more information.
+
+
```
```suggestion
```

NielsRogge commented on src/transformers/models/vivit/configuration_vivit.py:
```diff
@@ -0,0 +1,119 @@
+
+class VivitConfig(PretrainedConfig):
+    r"""
+    This is the configuration class to store the configuration of a [`VivitModel`]. It is used to instantiate a Vivit
+    model according to the specified arguments, defining the model architecture. Instantiating a configuration with the
+    defaults will yield a similar configuration to that of the Vivit
+    [jegormeister/vivit-b-16x2-kinetics400](https://huggingface.co/jegormeister/vivit-b-16x2-kinetics400) architecture.
```
```suggestion
    This is the configuration class to store the configuration of a [`VivitModel`]. It is used to instantiate a ViViT
    model according to the specified arguments, defining the model architecture. Instantiating a configuration with the
    defaults will yield a similar configuration to that of the ViViT
    [google/vivit-b-16x2-kinetics400](https://huggingface.co/google/vivit-b-16x2-kinetics400) architecture.
```

NielsRogge commented on src/transformers/models/vivit/configuration_vivit.py:
````diff
@@ -0,0 +1,119 @@
+        Example:
+
+    ```python
+    >>> from transformers import VivitModel, VivitConfig
````
Please use the new code snippets format here, see bert as an example

jegork replied:
Could you please point me to the exact place that you are mentioning? As far as I see here https://github.com/huggingface/transformers/blob/main/src/transformers/models/bert/configuration_bert.py for `BertConfig`, the format is the same

NielsRogge replied:
Yes as can be seen, the order of imports is different, and the comments are also different

NielsRogge commented on src/transformers/models/vivit/configuration_vivit.py:
```diff
@@ -0,0 +1,119 @@
+        qkv_bias (`bool`, *optional*, defaults to `True`):
+            Whether to add a bias to the queries, keys and values.
+
+        Example:
```
```suggestion
Example:
```

NielsRogge commented on src/transformers/models/vivit/convert_vivit_flax_to_pytorch.py:
```diff
@@ -0,0 +1,134 @@
+"""Convert Flax checkpoints from original paper to PyTorch"""
```
```suggestion
# coding=utf-8
# Copyright 2022 The HuggingFace Inc. team.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Convert Flax ViViT checkpoints from original repository to PyTorch.

URL: https://github.com/google-research/scenic/tree/main/scenic/projects/vivit
"""
```

NielsRogge commented on src/transformers/models/vivit/convert_vivit_flax_to_pytorch.py:
```diff
@@ -0,0 +1,134 @@
+    return sum([1 if "encoderblock_" in k else 0 for k in state_dict["optimizer"]["target"]["Transformer"].keys()])
+
+
+if __name__ == "__main__":
```
The conversion script doesn't seem to verify logits/hidden states against the original implementation on the same input data. 

See VideoMAE or X-CLIP's conversion scripts as examples

NielsRogge commented on src/transformers/models/vivit/modeling_vivit.py:
```diff
@@ -0,0 +1,746 @@
+# coding=utf-8
+# Copyright 2022 Anurag Arnab, Mostafa Dehghani, Georg Heigold, Chen Sun, Mario Lučić, Cordelia Schmid The HuggingFace Inc. team. All rights reserved.
```
```suggestion
# Copyright 2022 Google AI and The HuggingFace Inc. team. All rights reserved.
```

NielsRogge commented on src/transformers/models/vivit/modeling_vivit.py:
```diff
@@ -0,0 +1,746 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" PyTorch Vivit model."""
```
```suggestion
""" PyTorch ViViT model."""
```

NielsRogge commented on src/transformers/models/vivit/modeling_vivit.py:
```diff
@@ -0,0 +1,746 @@
+
+logger = logging.get_logger(__name__)
+
+_CHECKPOINT_FOR_DOC = "jegormeister/vivit-b-16x2-kinetics400"
```
```suggestion
_CHECKPOINT_FOR_DOC = "google/vivit-b-16x2-kinetics400"
```

NielsRogge commented on src/transformers/models/vivit/modeling_vivit.py:
```diff
@@ -0,0 +1,746 @@
+]
+
+
+def load_tf_weights_in_vivit(model, config, tf_checkpoint_path):
```
This function can be deleted, unless you're using it (but I guess it's not used to convert weights)

NielsRogge commented on src/transformers/models/vivit/modeling_vivit.py:
```diff
@@ -0,0 +1,746 @@
+    return model
+
+
+class TubeletEmbeddings(nn.Module):
```
```suggestion
class VivitTubeletEmbeddings(nn.Module):
```

NielsRogge commented on src/transformers/models/vivit/modeling_vivit.py:
```diff
@@ -0,0 +1,746 @@
+
+
+@add_start_docstrings(
+    "The bare Vivit Model transformer outputting raw hidden-states without any specific head on top.",
```
```suggestion
    "The bare ViViT Transformer model outputting raw hidden-states without any specific head on top.",
```

NielsRogge commented on src/transformers/models/vivit/modeling_vivit.py:
```diff
@@ -0,0 +1,746 @@
+
+    def _prune_heads(self, heads_to_prune):
+        """
+        Prunes heads of the model. heads_to_prune: dict of {layer_num: list of heads to prune in this layer} See base
```
Docstring here seems incorrect

jegork replied:
fixed, but it seems the same formatting problem appears in many models (at least vit and videomae, bert)

NielsRogge commented on src/transformers/models/vivit/modeling_vivit.py:
```diff
@@ -0,0 +1,746 @@
+
+
+@add_start_docstrings(
+    """Vivit Model transformer with a video classification head on top (a linear layer on top of the final hidden state of
+    the [CLS] token) e.g. for ImageNet.""",
```
```suggestion
    """ViViT Transformer model with a video classification head on top (a linear layer on top of the final hidden state of the [CLS] token) e.g. for Kinetics-400.""",
```

NielsRogge commented on src/transformers/models/vivit/modeling_vivit.py:
```diff
@@ -0,0 +1,746 @@
+_TOKENIZER_FOR_DOC = "VivitTokenizer"
+
+VIVIT_PRETRAINED_MODEL_ARCHIVE_LIST = [
+    "jegormeister/vivit-b-16x2-kinetics400",
```
Same here

NielsRogge commented on src/transformers/models/vivit/modeling_vivit.py:
```diff
@@ -0,0 +1,746 @@
+
+_CHECKPOINT_FOR_DOC = "jegormeister/vivit-b-16x2-kinetics400"
+_CONFIG_FOR_DOC = "VivitConfig"
+_TOKENIZER_FOR_DOC = "VivitTokenizer"
```
Vivit doesn't have a tokenizer

NielsRogge commented on src/transformers/models/vivit/modeling_vivit.py:
```diff
@@ -0,0 +1,746 @@
+    (width // tubelet_size[2]).
+    """
+
+    def __init__(self, video_size, tubelet_size, num_channels=3, embed_dim=768):
```
For consistency with other models like ViT and VideoMAE, it'd be great to have this class just take a `config` argument, and use its attributes inside the class.

See [here](https://github.com/huggingface/transformers/blob/b210c83a78022226ce48402cd67d8c8da7afbd8d/src/transformers/models/videomae/modeling_videomae.py#L162-L179) as an example

NielsRogge commented on src/transformers/models/vivit/configuration_vivit.py:
```diff
@@ -0,0 +1,119 @@
+
+
+    Args:
+        video_size (`List[int]`, *optional*, defaults to [32, 224, 224]):
+            The size (resolution) of a video in the format `[number_of_frames, width, height]`.
+        tubelet_size (`List[int]`, *optional*, defaults to [2, 16, 16]):
+            The size (resolution) of each tubelet.
```
(nit) for consistency with other models like ViT and VideoMAE, it'd be great to have a `num_frames` attribute in the config, and an `image_size` attribute (rather than video_size). The `image_size` can be an integer or a tuple.

See also my comment in the `TubeletEmbeddings` class

NielsRogge commented on tests/models/vivit/test_modeling_vivit.py:
```diff
@@ -0,0 +1,359 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" Testing suite for the PyTorch Vivit model. """
```
```suggestion
""" Testing suite for the PyTorch ViViT model. """
```

NielsRogge commented on tests/models/vivit/test_modeling_vivit.py:
```diff
@@ -0,0 +1,359 @@
+
+    @slow
+    def test_inference_for_video_classification(self):
+        model = VivitForVideoClassification.from_pretrained("jegormeister/vivit-b-16x2-kinetics400").to(torch_device)
```
To be updated

NielsRogge reviewed (commented):
PR looks in great shape already. Main comments are:

- verification of the logits/hidden states against original implementation in the conversion script
- would also be nice to make sure the image processor is using the exact same transformations as the original implementation (things like resizing the shortest edge + center cropping, which interpolation method is used, which image mean + std,...) => this can also be verified in the conversion script

jegork committed:
Fix docstring and formatting

jegork committed:
Rename TubeletEmbeddings -> VivitTubeletEmbeddings

jegork committed:
Remove load_tf_weights_in_vivit

jegork committed:
Change checkpoint name

jegork committed:
Remove Vivit _TOKENIZER_FOR_DOC

jegork committed:
Fix

jegork committed:
Fix VivitTubeletEmbeddings and pass config object as parameter

jegork committed:
Use image_size and num_frames instead of video_size

jegork commented:
hey @NielsRogge, thank you for your comments!

I have a question regarding the conversion script. I am using `restore_checkpoint(flax_model_path, None)` from `flax.training.checkpoints`, however, using the current jax version 0.3.6 (it was installed on my machine automatically when installing dev transformers version as per documentation). Upgrading the version of flax to 0.3.25 fixes the issue (haven't tested lower versions). Is there any other way I could load a flax checkpoint?

jegork committed:
Change conversion script and fix differences with the orig implementation

jegork committed:
Fix docstrings

jegork committed:
Add attention head pruning

jegork committed:
Run style and fixup

jegork committed:
Fix tests

jegork committed:
Add ViViT to video_classification.mdx

jegork commented:
@NielsRogge i think I've addressed all of your points, though I have doubt whether I implemented correctly the transformations check in the conversion script (I couldn't come up with any better approach as the preprocessing code in the original implementation is kinda scattered among multiple files, but I have tried to leave as much as possible of notes there). Would be great if you could check that and, if applicable, suggest any improvements.

Thanks again!

jegork committed:
Save processor in conversion script

jegork commented:
seems like the test job fails with
```
FAILED tests/models/bridgetower/test_modeling_bridgetower.py::BridgeTowerModelTest::test_save_load_fast_init_to_base - AssertionError: 0.08835485577583313 not less than or equal to 0.001 : vision_model.visual.transformer.resblocks.11.attn.in_proj_weight not identical
```

after i rebased from the main branch. Seems like has nothing to do with this PR. Was it already fixed and shall I rebase again?

NielsRogge commented:
Sorry for the late reply here, I've assigned @amyeroberts to review the PR.

amyeroberts commented on src/transformers/models/vivit/image_processing_vivit.py:
```diff
@@ -0,0 +1,385 @@
+        Preprocess an image or batch of images.
+
+        Args:
+            images (`ImageInput`):
```
```suggestion
            videos (`ImageInput`):
```

amyeroberts commented on src/transformers/models/vivit/image_processing_vivit.py:
```diff
@@ -0,0 +1,385 @@
+                Whether to centre crop the image.
+            crop_size (`Dict[str, int]`, *optional*, defaults to `self.crop_size`):
+                Size of the image after applying the centre crop.
+            do_rescale (`bool`, *optional*, defaults to `self.do_rescale`):
+                Whether to rescale the image values between [0 - 1].
```
As `do_zero_centering` is by default `True` - I would indicate that in the output description here 
```suggestion
            do_rescale (`bool`, *optional*, defaults to `self.do_rescale`):
                Whether to rescale the image values between `[-1 - 1]` if `do_zero_centering` is `True`, `[0, 1]` otherwise.
```

amyeroberts commented on src/transformers/models/vivit/image_processing_vivit.py:
```diff
@@ -0,0 +1,385 @@
+            raise ValueError(f"Size must have 'height' and 'width' as keys. Got {size.keys()}")
+        return center_crop(image, size=(size["height"], size["width"]), data_format=data_format, **kwargs)
+
+    def rescale(
```
Rather than `do_center_zeroing`, can you match the pattern in the efficientnet image processor [that uses an offset parameter](https://github.com/huggingface/transformers/blob/7811bf7e736f20d0990d3630ae33e42e373b8e86/src/transformers/models/efficientnet/image_processing_efficientnet.py#L178)?

amyeroberts commented on tests/models/vivit/test_modeling_vivit.py:
```diff
@@ -0,0 +1,354 @@
+        qkv_bias=True,
+        scope=None,
+    ):
+        self.parent = parent
```
nit: could you order the definition of the instance attributes in the same order as the function signature  - it makes it easier to read and check i.e. 

```
def __init__(self, a, b, ...):
    self.a = a
    self.b = b
    ...
```

amyeroberts reviewed (commented):
Thanks for adding this model! 

PR is looking good. Just a few pieces which need to be added before we merge. Can you make sure to add:
* Copyright headers with up to date info in all the module files
* Tests for the image processors i.e. a `test_image_processing_vivit.py` file? 

Regarding the bridgeformer failing tests - I think your previous suggestion is a good idea: rebase from main - as these tests shouldn't be affect by this PR :)
